### PR TITLE
draft - cross account attachment resource

### PR DESCRIPTION
**Pull request by kathmbeck**
### Description

Draft for cross account attachment implementation 

### Relations

Closes #0000

### Output from Acceptance Testing
Passing acceptance tests 
```
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/globalaccelerator/... -v -count 1 -parallel 20 -run='TestAccGlobalAcceleratorCrossAccountAttachment_basic'  -timeout 360m
=== RUN   TestAccGlobalAcceleratorCrossAccountAttachment_basic
=== PAUSE TestAccGlobalAcceleratorCrossAccountAttachment_basic
=== CONT  TestAccGlobalAcceleratorCrossAccountAttachment_basic
--- PASS: TestAccGlobalAcceleratorCrossAccountAttachment_basic (13.74s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/globalaccelerator  19.684s

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/globalaccelerator/... -v -count 1 -parallel 20 -run='TestAccGlobalAcceleratorCrossAccountAttachment_principals'  -timeout 360m
=== RUN   TestAccGlobalAcceleratorCrossAccountAttachment_principals
=== PAUSE TestAccGlobalAcceleratorCrossAccountAttachment_principals
=== CONT  TestAccGlobalAcceleratorCrossAccountAttachment_principals
--- PASS: TestAccGlobalAcceleratorCrossAccountAttachment_principals (13.46s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/globalaccelerator  19.293s

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/globalaccelerator/... -v -count 1 -parallel 20 -run='TestAccGlobalAcceleratorCrossAccountAttachment_resources'  -timeout 360m
=== RUN   TestAccGlobalAcceleratorCrossAccountAttachment_resources
=== PAUSE TestAccGlobalAcceleratorCrossAccountAttachment_resources
=== CONT  TestAccGlobalAcceleratorCrossAccountAttachment_resources
--- PASS: TestAccGlobalAcceleratorCrossAccountAttachment_resources (18.48s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/globalaccelerator  24.444s

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/globalaccelerator/... -v -count 1 -parallel 20 -run='TestAccGlobalAcceleratorCrossAccountAttachment_disappears'  -timeout 360m
=== RUN   TestAccGlobalAcceleratorCrossAccountAttachment_disappears
=== PAUSE TestAccGlobalAcceleratorCrossAccountAttachment_disappears
=== CONT  TestAccGlobalAcceleratorCrossAccountAttachment_disappears
--- PASS: TestAccGlobalAcceleratorCrossAccountAttachment_disappears (12.10s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/globalaccelerator  18.071s
```

### CODE REVIEW

File: internal/service/globalaccelerator/cross_account_attachment.go [hunk 1]
```diff
@@ -0,0 +1,534 @@
...
+					listplanmodifier.RequiresReplace(),
+				},
+			},
+			"resources": schema.ListAttribute{
+				Optional:    true,
+				ElementType: ResourceDataElementType,
+				PlanModifiers: []planmodifier.List{
+					listplanmodifier.RequiresReplace(),
+				},
+			},
```
Comment (mraerino) [inline internal/service/globalaccelerator/cross_account_attachment.go]: it looks like you're able to update these in-place. why are they marked as `RequiresReplace`?

File: website/docs/r/globalaccelerator_cross_account_attachment.html.markdown [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+## Argument Reference
+
+The following arguments are required:
+
+* `name` - (Required) Name of the Cross Account Attachment.
+
+The following arguments are optional:
+
+* `principals` - (Optional) List of AWS account IDs that are allowed to associate resources with the accelerator.
+* `resources` - (Optional) List of resources to be associated with the accelerator. Each resource is specified as a map with keys `endpoint_id` and `region.
```
Comment (mraerino) [inline website/docs/r/globalaccelerator_cross_account_attachment.html.markdown]: nit: missing end quote
```suggestion
* `resources` - (Optional) List of resources to be associated with the accelerator. Each resource is specified as a map with keys `endpoint_id` and `region`.
```

File: website/docs/r/globalaccelerator_cross_account_attachment.html.markdown [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+## Attribute Reference
+
+This resource exports the following attributes in addition to the arguments above:
+
+* `arn` - ARN of the Cross Account Attachment. 
+* `id` - ID of the Cross Account Attachment.
+* `created_time` - Creation Time when the Cross Account Attachment.
+* `last_modified_time` - Last modified time of the Cross Account Attachment. 
+*
```
Comment (mraerino) [inline website/docs/r/globalaccelerator_cross_account_attachment.html.markdown]: nit: superfluous bullet

File: website/docs/r/globalaccelerator_cross_account_attachment.html.markdown [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+* `delete` - (Default `30m`)
+
+## Import
+
+In Terraform v1.5.0 and later, use an [`import` block](https://developer.hashicorp.com/terraform/language/import) to import Global Accelerator Cross Account Attachment using the `example_id_arg`. For example:
+
+```terraform
+import {
+  to = aws_globalaccelerator_cross_account_attachment.example
+  id = "cross_account_attachment-id-12345678"
```
Comment (mraerino) [inline website/docs/r/globalaccelerator_cross_account_attachment.html.markdown]: isn't the `id` supposed to be an ARN? seems so from the code?

File: internal/service/globalaccelerator/cross_account_attachment.go [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+
+	elems := []attr.Value{}
+	for _, resource := range resources {
+		endpointID := aws.StringValue(resource.EndpointId)
+		region := ""
+		// Extract the region from the ARN if the endpoint ID is an ARN
+		if strings.HasPrefix(endpointID, "arn:") {
+			parts := strings.Split(endpointID, ":")
+			if len(parts) > 3 {
+				region = parts[3]
```
Comment (mraerino) [inline internal/service/globalaccelerator/cross_account_attachment.go]: the AWS SDK offers functions to [detect](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2@v1.25.2/aws/arn#IsARN) and [parse](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2@v1.25.2/aws/arn#Parse) an ARN
feels like a proper parser provides more reliability for this?

File: internal/service/globalaccelerator/cross_account_attachment.go [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+	if diags.HasError() {
+		return
+	}
+}
+
+var ResourceDataElementType = types.ObjectType{
+	AttrTypes: map[string]attr.Type{
+		"endpoint_id": types.StringType,
+		"region":      types.StringType,
+	},
```
Comment (mraerino) [inline internal/service/globalaccelerator/cross_account_attachment.go:307]: is there anything that needs to be done here to mark `region` as optional?

File: internal/service/globalaccelerator/cross_account_attachment.go [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+	}
+	if out == nil || out.CrossAccountAttachment == nil {
+		resp.Diagnostics.AddError(
+			create.ProblemStandardMessage(names.GlobalAccelerator, create.ErrActionCreating, ResNameCrossAccountAttachment, plan.Name.String(), nil),
+			errors.New("empty output").Error(),
+		)
+		return
+	}
+
+	plan.ID = flex.StringToFramework(ctx, out.CrossAccountAttachment.AttachmentArn)
```
Comment (mraerino) [inline internal/service/globalaccelerator/cross_account_attachment.go]: why is the `ID` assigned to `plan` here? i don't see `plan` used later? could it just be assigned to `state`?

File: internal/service/globalaccelerator/cross_account_attachment.go [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+	if out.CrossAccountAttachment.LastModifiedTime != nil {
+		state.LastModifiedTime = types.StringValue(out.CrossAccountAttachment.LastModifiedTime.Format(time.RFC3339))
+	}
+
+	resources, errDiags := flattenResources(ctx, out.CrossAccountAttachment.Resources)
+	resp.Diagnostics.Append(errDiags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	state.Resources = resources
```
Comment (mraerino) [inline internal/service/globalaccelerator/cross_account_attachment.go:182]: why are you setting the `Resources` but not the `Principals`?

File: internal/service/globalaccelerator/cross_account_attachment.go (hunk 1 above)
Comment (kathmbeck) [reply internal/service/globalaccelerator/cross_account_attachment.go]: Good callout. I'll remove

File: internal/service/globalaccelerator/cross_account_attachment.go (hunk 6 above)
Comment (kathmbeck) [reply internal/service/globalaccelerator/cross_account_attachment.go:307]: I don't believe so, I'm treating it as optional below in `expandResources`, as well as handling the case where its not present in `flattenResources`. I think adding a note to the documentation page might be beneficial.

File: internal/service/globalaccelerator/cross_account_attachment.go (hunk 5 above)
Comment (kathmbeck) [reply internal/service/globalaccelerator/cross_account_attachment.go]: Oh nice 💯

File: internal/service/globalaccelerator/cross_account_attachment.go (hunk 6 above)
Comment (kathmbeck) [reply internal/service/globalaccelerator/cross_account_attachment.go:307]: Added a note in the latest commit.

File: internal/service/globalaccelerator/cross_account_attachment.go (hunk 8 above)
Comment (kathmbeck) [reply internal/service/globalaccelerator/cross_account_attachment.go:182]: Added setting principals.

**Comment by kathmbeck** [comment]
Thanks for the quick review @mraerino 😄

**Review by mraerino — COMMENTED** [review]
nice, looks good!

did you want to test this with the real APIs? i think there are ways to manually compile and run locally, if you have the AWS creds

**Comment by kathmbeck** [comment]
> nice, looks good!
> 
> did you want to test this with the real APIs? i think there are ways to manually compile and run locally, if you have the AWS creds

I did, the acceptance tests run/pass against our sandbox account!

**Comment by coolfrood** [comment]
@ewbankkit  I'm on AWS Global Accelerator team. One of our partners has contributed this PR to provide support for a new resource that is supported by Global Accelerator. Could you please take a look? Thanks!

**Comment by kathmbeck** [comment]
WIP draft

**Comment by kathmbeck** [comment]
> @ewbankkit I'm on AWS Global Accelerator team. One of our partners has contributed this PR to provide support for a new resource that is supported by Global Accelerator. Could you please take a look? Thanks!

@coolfrood Just to clarify, here is the PR that needs the review. https://github.com/hashicorp/terraform-provider-aws/pull/35991

**Pull request closed without merging**